### PR TITLE
Add some new loss functions

### PR DESCRIPTION
**Pull request by ahwillia**
**Work In Progress - Don't Merge Yet**

To do:

- [x] `PeriodicLoss` - useful for estimating parameters on a circular scale.
- [x] `PoissonLoss`
- [x] `HuberLoss`
- [ ] ~~`OrdinalLoss` - http://ttic.uchicago.edu/~nati/Publications/RennieSrebroIJCAI05.pd~~ Left for future PR

### CODE REVIEW

**Comment by Evizero** [comment]
It would be nice if you could also add a test against a - as simple as possible - reference function similar to this https://github.com/JuliaML/Losses.jl/blob/master/test/tst_loss.jl#L232-L238

File: src/supervised/distance.jl [hunk 1]
```diff
@@ -108,6 +108,40 @@ islipschitzcont_deriv(::L2DistLoss) = true
...
+immutable PeriodicLoss{T<:AbstractFloat} <: DistanceLoss
+    k::T   # k = 2π/circumference
+    PeriodicLoss(circ::T) = new(T(2π/circ))
+end
+PeriodicLoss{T<:AbstractFloat}(circ::T=1.0) = PeriodicLoss{T}(circ)
+PeriodicLoss{T<:Number}(circ::T) = PeriodicLoss{Float64}(Float64(circ))
+
+value{T<:Number}(loss::PeriodicLoss, difference::T) = 1 - cos(difference*loss.k)
+deriv{T<:Number}(loss::PeriodicLoss, difference::T) = loss.k * sin(difference*loss.k)
+deriv2{T<:Number}(loss::PeriodicLoss, difference::T) = loss.k^2 * cos(difference*loss.k)
```
Comment (Evizero) [inline src/supervised/distance.jl]: I don't think it makes too much of a difference, but i always use `abs2(x)` instead of `x^2`. I think the special case of ^2 is a little more optimized

File: src/supervised/distance.jl [hunk 2]
```diff
@@ -108,6 +108,40 @@ islipschitzcont_deriv(::L2DistLoss) = true
...
+# L(y, t) = 1 - cos((y-t)*2π/c)
+
+"""
+`PeriodicLoss <: DistanceLoss`
+
+Measures distance on a circle of specified circumference.
+""" 
+immutable PeriodicLoss{T<:AbstractFloat} <: DistanceLoss
+    k::T   # k = 2π/circumference
+    PeriodicLoss(circ::T) = new(T(2π/circ))
```
Comment (Evizero) [inline src/supervised/distance.jl]: Does the loss make sense with negative circ? It doesn't look like it would cause numerical issues. But I would at least check for `0` like here: https://github.com/JuliaML/Losses.jl/blob/master/src/supervised/distance.jl#L118 ?

Comment (Evizero) [inline src/supervised/distance.jl]: Then again I suppose if one specifies a zero an exception is thrown anyway. so maybe this is actually fine as it is (if a negative `circ` is allows)

Comment (ahwillia) [inline src/supervised/distance.jl]: Good catch! Not sure what to do about negative circumference - since cosine is an even function it is sort of valid and if you accidentally do `PeriodicLoss(-5)` it will behave the same as `PeriodicLoss(5)`. But from a semantics/mathematical perspective a negative circumference doesn't make sense. I'm going to throw an error, unless you think it is better not to interfere.

Comment (Evizero) [inline src/supervised/distance.jl]: I leave this to your judgement.

**Comment by ahwillia** [comment]
@Evizero - for float-forcing type stability, do we want to allow `Float32` to pass through without changing? This appears to be violated all over the place...

To be concrete, I'm asking whether [this test](https://github.com/JuliaML/Losses.jl/blob/master/test/tst_loss.jl#L17) should be changed to:

```julia
if typeof(t) <: Float32
    @test (typeof(val) <: Float32)
else
    @test (typeof(val) <: Float64)
end
```

Thanks for walking me through this!

**Comment by Evizero** [comment]
Yes, that was my intention. I went through the code to see why I did the test like that (because it must have been on purpose, even if it was a stupid reason) and I think it was because I made the decision to force Loss parameters to Float64 like here : https://github.com/JuliaML/Losses.jl/blob/master/src/supervised/distance.jl#L115

Now I can't think of a reason why I would want that to be fixed (Can you?). Maybe we should change them to be dynamic (like yours currently is) and fix the tests as you proposed

**Comment by Evizero** [comment]
I don't think splitting the if would fix the test logic though. We would have to split the test loop into two loops. one for Float32 and one for Float64 (because it would effectively depend on how the Loss is created (`PeriodicLoss(1.0)` vs `PeriodicLoss(Float32(1.0))`)

File: src/supervised/distance.jl [hunk 3]
```diff
@@ -126,10 +126,10 @@ PeriodicLoss{T<:Number}(circ::T) = PeriodicLoss{Float64}(Float64(circ))
 
 value{T<:Number}(loss::PeriodicLoss, difference::T) = 1 - cos(difference*loss.k)
 deriv{T<:Number}(loss::PeriodicLoss, difference::T) = loss.k * sin(difference*loss.k)
-deriv2{T<:Number}(loss::PeriodicLoss, difference::T) = loss.k^2 * cos(difference*loss.k)
+deriv2{T<:Number}(loss::PeriodicLoss, difference::T) = abs2(loss.k) * cos(difference*loss.k)
 function value_deriv{T<:Number}(loss::PeriodicLoss, difference::T)
     dk = difference*loss.k
-    return 1-cos(dk), T(loss.k)*sin(dk)
+    return 1-cos(dk), loss.k*sin(dk)
```
Comment (Evizero) [inline src/supervised/distance.jl]: I like this change. Now we leave it up to appropriate type promotion. 👍

**Comment by ahwillia** [comment]
Yeah I think the most flexible way to do it would be like:

```julia
immutable L1EpsilonInsLoss{T<:AbstractFloat} <: DistanceLoss
    eps::T
end
```

I don't use `Float32` but I guess people might... Sigh. I can take care of this under this PR if you give me the 👍 

> I don't think splitting the if would fix the test logic though. We would have to split the test loop into two loops. one for Float32 and one for Float64

Ah, I think I wanted:

```julia
T = promote_op(typeof(y),typeof(t))
if T <: Float32
    @test (typeof(val) <: Float32)
else
    @test (typeof(val) <: Float64)
end
```

Why do we need two loops?

**Comment by Evizero** [comment]
> Why do we need two loops?

Because `T` also depends on the type of a possible field of `l`, such as `circ`

**Comment by Evizero** [comment]
I was thinking maybe two of https://github.com/JuliaML/Losses.jl/blob/master/test/tst_loss.jl#L175 instead of one. One time for `Float64 forcing` (which tests `L1EpsilonInsLoss(0.5)`, etc)  and one time for `Float32 preserving` (which tests `L1EpsilonInsLoss(Float32(0.5))`, `LogitMarginLoss()`, etc)

**Comment by Evizero** [comment]
> Thanks for walking me through this!

Any time! I am really glad that this turned into a real community package and I will help where I can. Thanks for working on this! (it is midnight soon though, so don't be surprised when I am suddenly gone soon)

File: src/supervised/distance.jl [hunk 4]
```diff
@@ -145,29 +153,43 @@ isstronglyconvex(::PeriodicLoss) = false
 # ===========================================================
 # L(y, t) = max(0, |y - t| - ɛ)
 
-immutable L1EpsilonInsLoss <: DistanceLoss
-    eps::Float64
+immutable L1EpsilonInsLoss{T<:AbstractFloat} <: DistanceLoss
+    ε::T
```
Comment (Evizero) [inline src/supervised/distance.jl]: Personally I think we should avoid unicode characters in anything "userfacing". But I could be convinced. Thoughts @tbreloff @joshday ?

File: src/supervised/scaledloss.jl [hunk 5]
```diff
@@ -1,4 +1,4 @@
-for KIND in (:MarginLoss, :DistanceLoss)
+for KIND in (:MarginLoss, :DistanceLoss, :SupervisedLoss)
```
Comment (Evizero) [inline src/supervised/scaledloss.jl]: a few lines down there is a `typealias` for `ScaledLoss` which should reflect this addition

File: test/tst_loss.jl [hunk 6]
```diff
@@ -2,25 +2,47 @@ function test_value_typestable(l::SupervisedLoss)
...
+                if !(T <: AbstractFloat)
+                    # cast Integers to a float
+                    # (whether its Float32 or Float64 depends on the loss...)
+                    @test (typeof(val) <: AbstractFloat)
+                elseif T <: Float32
+                    # preserve Float32
+                    @test (typeof(val) <: Float32)
+                else
+                    @test (typeof(val) <: Float64)
+                end
```
Comment (Evizero) [inline test/tst_loss.jl]: nice way to solve this!

File: test/tst_loss.jl [hunk 7]
```diff
@@ -172,21 +194,29 @@ end
 
 @testset "Test typestable supervised loss for type stability" begin
     for loss in [L1HingeLoss(), L2HingeLoss(), ModifiedHuberLoss(), PerceptronLoss(),
-                ZeroOneLoss(), LPDistLoss(1), LPDistLoss(2), LPDistLoss(3)]
+                LPDistLoss(1), LPDistLoss(2), LPDistLoss(3)]
         test_value_typestable(loss)
-        if typeof(loss) <: Union{MarginLoss, DistanceLoss}
-            test_value_floatforcing(2. * loss)
-        end
+        # TODO: add ZeroOneLoss after scaling works...
```
Comment (Evizero) [inline test/tst_loss.jl]: What happened that broke this?

File: test/tst_loss.jl [hunk 8]
```diff
@@ -228,41 +258,43 @@ end
 end
 
 @testset "Test distance-based loss against reference function" begin
+    yr,tr = linspace(-20,20,10),linspace(-30,30,10)
```
Comment (Evizero) [inline test/tst_loss.jl]: good idea. never occurred to me...

File: src/supervised/distance.jl (hunk 4 above)
Comment (ahwillia) [inline src/supervised/distance.jl]: Do pings work in comments? I seem to remember someone having problems with this.

I believe users should be comfortable with unicode, this is what Distributions.jl does. When I was first starting out using Julia it tripped me up for a little bit, I maybe got a little frustrated, but then I realized it was 100% the way to go and I was a fool for thinking otherwise.

My real reasoning for changing this was `eps` can be confused with `Base.eps()`. So if there is an alternative I can happily put that in.

File: src/supervised/scaledloss.jl (hunk 5 above)
Comment (ahwillia) [inline src/supervised/scaledloss.jl]: This was actually a mistake -- I tried this out and decided against it. I'll revert back in the next commit?

File: src/supervised/distance.jl (hunk 4 above)
Comment (Evizero) [inline src/supervised/distance.jl]: I can see that reasoning. I could go either way. Let's see what the others say.

File: src/supervised/scaledloss.jl (hunk 5 above)
Comment (ahwillia) [inline src/supervised/scaledloss.jl]: One note -- `ZeroOneLoss` does not work with `ScaledLoss` at the moment. How would you want to resolve this? Can't it be considered a `MarginLoss`?

File: test/tst_loss.jl (hunk 7 above)
Comment (ahwillia) [inline test/tst_loss.jl]: `test_value_typestable` now also tests a scaled loss, but this fails for `ZeroOneLoss` since it is not either a Distance or MarginLoss

File: src/supervised/scaledloss.jl (hunk 5 above)
Comment (Evizero) [inline src/supervised/scaledloss.jl]: ZeroOneLoss is not really a practical loss. It is the "ideal" classification goal of sorts. It is mostly in the package for completeness sake and so that users can plot it. AFAIK it is not a proper MarginLoss. Looking at the code now I don't remember why not. Need to look in the book again

File: test/tst_loss.jl (hunk 7 above)
Comment (Evizero) [inline test/tst_loss.jl]: yes, I remember. I will revisit the `ZeroOneLoss` sooner or later. This is perfectly fine for now

File: test/tst_loss.jl [hunk 9]
```diff
@@ -24,6 +52,8 @@ function test_value(l::SupervisedLoss, f::Function, y_vec, t_vec)
     @testset "$(l): " begin
         for y in y_vec, t in t_vec
             @test abs(value(l, y, t) - f(y, t)) < 1e-10
+            # TODO: consider this test?
+            # @test abs(value_deriv(l,y,t)[1] - f(y, t)) < 1e-10
```
Comment (Evizero) [inline test/tst_loss.jl]: this is covered by other tests. see: https://github.com/JuliaML/Losses.jl/blob/master/test/tst_loss.jl#L74

File: src/supervised/distance.jl [hunk 10]
```diff
@@ -108,32 +108,88 @@ islipschitzcont_deriv(::L2DistLoss) = true
...
+        new(convert(T, 2π/circ))
+    end
+end
+function PeriodicLoss{T<:Number}(ε::T=1.0)
+    if T <: AbstractFloat
+        PeriodicLoss{T}(ε)
+    else # cast to Float64
+        PeriodicLoss{Float64}(Float64(ε))
+    end
+end
```
Comment (Evizero) [inline src/supervised/distance.jl]: What was wrong with the two method approach? Did the compiler complain about ambiguity ?

```julia
PeriodicLoss{T<:AbstractFloat}(circ::T=1.0) = PeriodicLoss{T}(circ)
PeriodicLoss(circ) = PeriodicLoss{Float64}(Float64(circ))
``` 

looks a bit nicer

Comment (ahwillia) [inline src/supervised/distance.jl]: I thought this way was more explicit. But could revert back if you disagree.

What sort of compiler checks should I be doing?

Comment (Evizero) [inline src/supervised/distance.jl]: I think it would be cleaner if we do the `if` by dispatch (my example) instead of a runtime check. Don't think we need a compiler check (I thought you changed it because the old way cause warnings or something).

File: src/supervised/distance.jl [hunk 11]
```diff
@@ -108,32 +108,154 @@ islipschitzcont_deriv(::L2DistLoss) = true
...
+        new(d)
+    end
+end
+HuberLoss{T<:AbstractFloat}(d::T=1.0) = HuberLoss{T}(d)
+HuberLoss{T<:Number}(d::T) = HuberLoss{Float64}(Float64(d))
+
+function value{T1,T2<:Number}(loss::HuberLoss{T1}, difference::T2)
+    T = promote_type(T1,T2)
+    abs_diff = abs(difference)
+    if abs_diff < loss.d
```
Comment (Evizero) [inline src/supervised/distance.jl]: should this be a `<=`?

File: src/supervised/distance.jl [hunk 12]
```diff
@@ -108,32 +108,154 @@ islipschitzcont_deriv(::L2DistLoss) = true
...
+        der = T(difference)
+    else
+        val = (loss.d*abs_diff) - T(0.5)*abs2(loss.d)
+        der = loss.d*T(sign(difference))
+    end
+    return val,der
+end
+
+isdifferentiable(::HuberLoss) = false
+isdifferentiable(l::HuberLoss, at) = at != abs(l.d)
```
Comment (Evizero) [inline src/supervised/distance.jl]: Is it not once differentiable? I think it is. Looks that way anyway. We should double check this.

Comment (pkofod) [inline src/supervised/distance.jl]: Isn't differentiability of the Huber loss function at the origin one of its benefits? :)

Comment (Evizero) [inline src/supervised/distance.jl]: The question is more about transition between the two functions it is made up of (i.e. at `l.d`). But I think it is smooth and continuous, so approaching from both sides to that point arbitrarily close should result in the same derivative.

File: src/supervised/distance.jl (hunk 4 above)
Comment (Evizero) [inline src/supervised/distance.jl]: since Tom is also ok with the unicode form let's go with your change. 👍

File: src/supervised/distance.jl (hunk 12 above)
Comment (ahwillia) [inline src/supervised/distance.jl]: You're totally right @Evizero  - egg on my face. I think I was confused by twice-differentiability (which it is not at `l.d`?).

File: src/supervised/distance.jl (hunk 11 above)
Comment (ahwillia) [inline src/supervised/distance.jl]: Yes.

File: src/supervised/distance.jl (hunk 12 above)
Comment (Evizero) [inline src/supervised/distance.jl]: > I think I was confused by twice-differentiability (which it is not at l.d?)


screen shot 2016-08-11 at 19 56 36 2

doubt it 😄

Comment (ahwillia) [inline src/supervised/distance.jl]: Just checking - I was starting to doubt myself 😄

File: src/supervised/distance.jl [hunk 13]
```diff
@@ -108,32 +108,152 @@ islipschitzcont_deriv(::L2DistLoss) = true
...
+    end
+end
+HuberLoss{T<:AbstractFloat}(d::T=1.0) = HuberLoss{T}(d)
+HuberLoss(d) = HuberLoss{Float64}(Float64(d))
+
+function value{T1,T2<:Number}(loss::HuberLoss{T1}, difference::T2)
+    T = promote_type(T1,T2)
+    abs_diff = abs(difference)
+    if abs_diff <= loss.d
+        return T(0.5)*abs2(difference)   # quadratic
```
Comment (Evizero) [inline src/supervised/distance.jl]: the `return` statements aren't really needed

File: src/supervised/distance.jl [hunk 14]
```diff
@@ -108,32 +108,152 @@ islipschitzcont_deriv(::L2DistLoss) = true
...
+        return loss.d*T(sign(difference))   # linear
+    end
+end
+function deriv2{T1,T2<:Number}(loss::HuberLoss{T1}, difference::T2)
+    T = promote_type(T1,T2)
+    if abs(difference) <= loss.d
+        return one(T)   # quadratic
+    else
+        return zero(T)   # linear
+    end
```
Comment (Evizero) [inline src/supervised/distance.jl]: maybe `abs(difference) <= loss.d ? one(T) : zero(T)`

**Comment by Evizero** [comment]
This is very nice already. Big fan of you additions/changes

File: src/supervised/distance.jl (hunk 13 above)
Comment (ahwillia) [inline src/supervised/distance.jl]: Yes, this is a stylistic difference I have with a lot of the Julia community. I'm not always consistent about it though.

File: src/supervised/distance.jl (hunk 14 above)
Comment (ahwillia) [inline src/supervised/distance.jl]: Good idea. Done.

**Comment by ahwillia** [comment]
I'm going to leave `OrdinalLoss` for a future PR, since I need to focus on a few other things. @Evizero - I'm leaving this to you to review one last time and merge. I'm happy to squash commits or make further edits as you see fit.

Next I want to make all of this work on v0.5 - stay tuned for that.

**Comment by Evizero** [comment]
Do you want to add `OrdinalLoss` here, or later? I'm ok merging this as it is

**Comment by Evizero** [comment]
You beat me to it

**Comment by Evizero** [comment]
maybe squash the related commits (two periodic loss, and the two huber loss)

**Comment by ahwillia** [comment]
Done.

**Comment by Evizero** [comment]
Thanks! This is a really nice contribution!
